File: game/forest/forest_game.py
import time

import pygame
import random

from config import Config
from effect_type import EffectType
from effects.splat import Splat
from forest.forest_resources import ForestResources
from forest.playing import Playing
from forest.talking_after_hurt import TalkingAfterHurt
from forest.talking_game_over import TalkingGameOver
from forest.wait_intro import WaitIntro
from null_state import NullState
from phone_events import PhoneEvents


class ForestGame:

    def __init__(self, parent):
        self.parent = parent

        self.score = 0
        self.ended = False
        self.lives = 3

        self._state = WaitIntro(self)
        self._state.on_enter()

        self.parallax_pos = 0

        self.obstacles = self.generate_obstacles()
        self.sacks = self.generate_sacks()
        self.leaves = self.generate_leaves()
        self.effect_status = None
        self.effect_start = None

        pygame.mixer.Sound.play(ForestResources.sfx_bg_atmosphere, loops=-1)

    def process_events(self, phone_events: PhoneEvents):
        if self.effect_status == EffectType.INVERT and time.time() - self.effect_start > Config.EFFECT_DURATION_ORB:
            [phone_events.press_2, phone_events.press_8] = [phone_events.press_8, phone_events.press_2]

        next_state = self._state.process_events(phone_events)

        if self.effect_status == EffectType.SPLAT and time.time() - self.effect_start > Config.EFFECT_DURATION_SPLAT:
            self.effect_status = None
        if self.effect_status == EffectType.INVERT and time.time() - self.effect_start > Config.EFFECT_DURATION_INVERT:
            self.effect_status = None

        if next_state is not None:
            self._state.on_exit()
            self._state = next_state
            self._state.on_enter()
            if isinstance(next_state, NullState):
                self.end()

    def render(self, screen):
        self._state.render(screen)

        if self.effect_start:
            dt = time.time() - self.effect_start

            if dt < Config.EFFECT_DURATION_ORB:
                return

            if self.effect_status == EffectType.SPLAT:
                Splat.render_splat(screen, time.time() - self.effect_start)

            if self.effect_status == EffectType.INVERT and isinstance(self._state, Playing):
                screen.blit(ForestResources.inverted_arrows, (0, 0))

    def reduce_lives(self):
        self.lives -= 1
        if self.lives == 0:
            return TalkingGameOver(self)
        else:
            return TalkingAfterHurt(self)

    def external_effect(self, effect):
        if self.effect_status is not None:
            return
        self.effect_status = effect
        self.effect_start = time.time()

    def render_background(self, screen):
        hills_speed = 6 * Config.FOREST_BG_SPEED_MULTIPLIER
        trees_speed = 12 * Config.FOREST_BG_SPEED_MULTIPLIER
        grass_speed = 30 * Config.FOREST_BG_SPEED_MULTIPLIER
        hills_width = ForestResources.bg_hillsday[0].get_width()
        trees_width = ForestResources.bg_trees[0].get_width()
        grass_width = ForestResources.grass[0].get_width()
        ground_width = ForestResources.bg_ground[0].get_width()

        hills_pos = (self.new_mod(- self.parallax_pos * hills_speed, hills_width), 0)
        trees_pos = (self.new_mod(- self.parallax_pos * trees_speed, trees_width), -24)
        grass_pos = (self.new_mod(- self.parallax_pos * grass_speed, grass_width), 172)
        ground_pos = (self.new_mod(- self.parallax_pos * Config.FOREST_GROUND_SPEED, ground_width), 158)
        hills_pos_next = (hills_pos[0] + hills_width, hills_pos[1])
        tress_pos_next = (trees_pos[0] + trees_width, trees_pos[1])

        mountain_pos = 320 - 96 - (self.parallax_pos - Config.FOREST_MAX_TIME) * Config.FOREST_GROUND_SPEED

        screen.blit(ForestResources.bg_gradient, (0,0))
        screen.blit(ForestResources.bg_hillsday[0], hills_pos)
        screen.blit(ForestResources.bg_hillsday[0], hills_pos_next)

        screen.blit(ForestResources.bg_trees[0], trees_pos)
        screen.blit(ForestResources.bg_trees[0], tress_pos_next)

        for i in range(5):
            ground_x = ground_pos[0] + i * ground_width
            screen.blit(ForestResources.bg_ground[0], (ground_x, ground_pos[1]))

        for i in range(12):
            grass_x = grass_pos[0] + i * grass_width
            screen.blit(ForestResources.grass[0], (grass_x, grass_pos[1]))

        screen.blit(ForestResources.end_mountain[0], (mountain_pos, -16))

    def render_bottom(self, screen):
        screen.blit(ForestResources.scoreboard, (0, 184))
        for i in range(self.lives):
            screen.blit(ForestResources.hugo_lives[0], (i * 40 + 32, 188))

        x_score = 200
        y_score = 194
        x_space = 24
        thousands = self.score // 1000
        hundreds = (self.score - thousands * 1000) // 100
        tens = (self.score - thousands * 1000 - hundreds * 100) // 10
        ones = self.score - thousands * 1000 - hundreds * 100 - tens * 10

        screen.blit(ForestResources.score_numbers[0], (x_score + x_space * 0, y_score), self.get_spritesheet_area(thousands))
        screen.blit(ForestResources.score_numbers[0], (x_score + x_space * 1, y_score), self.get_spritesheet_area(hundreds))
        screen.blit(ForestResources.score_numbers[0], (x_score + x_space * 2, y_score), self.get_spritesheet_area(tens))
        screen.blit(ForestResources.score_numbers[0], (x_score + x_space * 3, y_score), self.get_spritesheet_area(ones))

    def end(self):
        self._state.on_exit()
        pygame.mixer.Sound.stop(ForestResources.sfx_bg_atmosphere)
        self.ended = True

    @staticmethod
    def get_spritesheet_area(value):
        width = 32
        height = 33

        xpos = value % 5
        ypos = value // 5

        return 1 + xpos * (width + 1), 1 + ypos * (height + 1), width, height

    @staticmethod
    def new_mod(a, b):
        res = a % b
        return res if not res else res - b if a < 0 else res

    @staticmethod
    def generate_obstacles():
        empty_prob = 0.65
        other_prob = (1 - empty_prob) / 4
        out = random.choices([0, 1, 2, 3, 4], weights=[empty_prob, other_prob, other_prob, other_prob, other_prob],
                             k=Config.FOREST_MAX_TIME)
        out[0] = 0
        out[1] = 0
        out[2] = 0
        out[3] = 0
        for idx in range(len(out) - 1):
            if out[idx] != 0:
                out[idx + 1] = 0

        return out

    @staticmethod
    def generate_sacks():
        empty_prob = 0.7
        other_prob = (1 - empty_prob) * 0.7
        other_prob_high = (1 - empty_prob) * 0.3
        return random.choices([0, 1, 2], weights=[empty_prob, other_prob, other_prob_high], k=Config.FOREST_MAX_TIME)

    @staticmethod
    def generate_leaves():
        arr = random.choices([1, 2], weights=[0.5, 0.5], k=Config.FOREST_MAX_TIME)
        for idx in range(len(arr) - 1):
            if arr[idx] == 2:
                arr[idx + 1] = 0
        return arr

File: game-web/src/forest/forestGame.ts
/** Port of game/forest/forest_game.py */

import { globalState } from '../globalState';
import { AudioHelper } from '../audioHelper';
import { Config } from '../config';
import { EffectType } from '../effectType';
import { ForestResources } from './forestResources';
import { WaitIntro } from './waitIntro';
import { Playing } from './playing';
import { NullState } from '../nullState';
import { Splat } from '../effects/splat';
import { RenderType } from '../renderType';
import { State, type StateClass } from '../state';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class ForestGame {
  context: GameData;
  ended = false;

  private _state: State;
  effectStatus: EffectType | null = null;
  effectStart: number | null = null;
  private forestBgAtmosphereStart: number;

  constructor(context: GameData) {
    this.context = context;
    context.forestScore = 0;
    context.forestLives = 3;
    context.forestParallaxPos = 0;
    context.forestObstacles = ForestGame.generateObstacles();
    context.forestSacks = ForestGame.generateSacks();
    context.forestLeaves = ForestGame.generateLeaves();

    this._state = new WaitIntro(context);
    this._state.onEnter();

    context.forestBgAtmosphereId = AudioHelper.play(ForestResources.sfxBgAtmosphere);
    this.forestBgAtmosphereStart = globalState.frameTime;
  }

  processEvents(phoneEvents: PhoneEvents): void {
    const invertActive =
      this.effectStatus === EffectType.INVERT &&
      globalState.frameTime - this.effectStart! > Config.EFFECT_DURATION_ORB;

    if (invertActive) {
      [phoneEvents.press_2, phoneEvents.press_8] = [phoneEvents.press_8, phoneEvents.press_2];
    }
    this.context.forestControlsInverted = invertActive;

    const nextStateClass = this._state.processEvents(phoneEvents);

    if (this.effectStatus === EffectType.SPLAT && globalState.frameTime - this.effectStart! > Config.EFFECT_DURATION_SPLAT) {
      this.effectStatus = null;
    }
    if (this.effectStatus === EffectType.INVERT && globalState.frameTime - this.effectStart! > Config.EFFECT_DURATION_INVERT) {
      this.effectStatus = null;
      this.context.forestControlsInverted = false;
    }

    if (nextStateClass !== null) {
      this._state.onExit();
      if (nextStateClass === NullState) {
        this.end();
      } else {
        this._state = new (nextStateClass as StateClass)(this.context);
        this._state.onEnter();
      }
    }

    if (!this.ended && globalState.frameTime - this.forestBgAtmosphereStart >= 5.5) {
      this.context.forestBgAtmosphereId = AudioHelper.play(ForestResources.sfxBgAtmosphere);
      this.forestBgAtmosphereStart = globalState.frameTime;
    }
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    const needsBg   = (this._state as any).needsBackground;
    const needsBot  = (this._state as any).needsBottom;

    if (needsBg  === RenderType.PRE) this.renderBackground(ctx);
    if (needsBot === RenderType.PRE) this.renderBottom(ctx);

    this._state.render(ctx);

    if (needsBg  === RenderType.POST) this.renderBackground(ctx);
    if (needsBot === RenderType.POST) this.renderBottom(ctx);

    if (this.effectStart !== null) {
      const dt = globalState.frameTime - this.effectStart;
      if (dt < Config.EFFECT_DURATION_ORB) return;

      if (this.effectStatus === EffectType.SPLAT) {
        Splat.renderSplat(ctx, globalState.frameTime - this.effectStart);
      }
      if (this.effectStatus === EffectType.INVERT && this._state instanceof Playing) {
        ctx.drawImage(ForestResources.invertedArrows!, 0, 0);
      }
    }
  }

  externalEffect(effect: EffectType): void {
    if (this.effectStatus !== null) return;
    this.effectStatus = effect;
    this.effectStart = globalState.frameTime;
  }

  private renderBackground(ctx: OffscreenCanvasRenderingContext2D): void {
    const hillsSpeed  = 6  * Config.FOREST_BG_SPEED_MULTIPLIER;
    const treesSpeed  = 12 * Config.FOREST_BG_SPEED_MULTIPLIER;
    const grassSpeed  = 30 * Config.FOREST_BG_SPEED_MULTIPLIER;

    const hillsWidth  = ForestResources.bgHillsday[0].width;
    const treesWidth  = ForestResources.bgTrees[0].width;
    const grassWidth  = ForestResources.grass[0].width;
    const groundWidth = ForestResources.bgGround[0].width;

    const hillsX  = ForestGame.newMod(-this.context.forestParallaxPos * hillsSpeed,  hillsWidth);
    const treesX  = ForestGame.newMod(-this.context.forestParallaxPos * treesSpeed,  treesWidth);
    const grassX  = ForestGame.newMod(-this.context.forestParallaxPos * grassSpeed,  grassWidth);
    const groundX = ForestGame.newMod(-this.context.forestParallaxPos * Config.FOREST_GROUND_SPEED, groundWidth);

    const mountainPos = 320 - 96 - (this.context.forestParallaxPos - Config.FOREST_MAX_TIME) * Config.FOREST_GROUND_SPEED;

    ctx.drawImage(ForestResources.bgGradient!, 0, 0);
    ctx.drawImage(ForestResources.bgHillsday[0], hillsX, 0);
    ctx.drawImage(ForestResources.bgHillsday[0], hillsX + hillsWidth, 0);
    ctx.drawImage(ForestResources.bgTrees[0], treesX, -24);
    ctx.drawImage(ForestResources.bgTrees[0], treesX + treesWidth, -24);

    for (let i = 0; i < 5; i++) ctx.drawImage(ForestResources.bgGround[0], groundX + i * groundWidth, 158);
    for (let i = 0; i < 12; i++) ctx.drawImage(ForestResources.grass[0], grassX + i * grassWidth, 172);

    ctx.drawImage(ForestResources.endMountain[0], mountainPos, -16);
  }

  private renderBottom(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(ForestResources.scoreboard!, 0, 184);
    for (let i = 0; i < this.context.forestLives; i++) {
      ctx.drawImage(ForestResources.hugoLives[0], i * 40 + 32, 188);
    }

    const xScoreRight = 150 + 24 * 6;
    const yScore = 194;
    const xSpace = 24;
    const score = Math.floor(this.context.forestScore);

    if (score === 0) {
      const [sx, sy, sw, sh] = ForestGame.getSpritesheetArea(0);
      ctx.drawImage(ForestResources.scoreNumbers[0], sx, sy, sw, sh, xScoreRight - xSpace, yScore, sw, sh);
      return;
    }

    const digits: number[] = [];
    let temp = score;
    while (temp > 0) { digits.push(temp % 10); temp = Math.floor(temp / 10); }

    for (let i = 0; i < digits.length; i++) {
      const [sx, sy, sw, sh] = ForestGame.getSpritesheetArea(digits[i]);
      ctx.drawImage(ForestResources.scoreNumbers[0], sx, sy, sw, sh, xScoreRight - xSpace * (i + 1), yScore, sw, sh);
    }
  }

  end(): void {
    this._state.onExit();
    AudioHelper.stop(this.context.forestBgAtmosphereId);
    this.ended = true;
  }

  static getSpritesheetArea(value: number): [number, number, number, number] {
    const width = 32, height = 33;
    const xpos = value % 5;
    const ypos = Math.floor(value / 5);
    return [1 + xpos * (width + 1), 1 + ypos * (height + 1), width, height];
  }

  static newMod(a: number, b: number): number {
    const res = a % b;
    if (!res) return res;
    return a < 0 ? res - b : res;
  }

  static generateObstacles(): number[] {
    const emptyProb = 0.65;
    const otherProb = (1 - emptyProb) / 4;
    const weights = [emptyProb, otherProb, otherProb, otherProb, otherProb];
    const out = weightedChoices([0, 1, 2, 3, 4], weights, Config.FOREST_MAX_TIME);
    out[0] = 0; out[1] = 0; out[2] = 0; out[3] = 0;
    for (let i = 0; i < out.length - 1; i++) if (out[i] !== 0) out[i + 1] = 0;
    return out;
  }

  static generateSacks(): number[] {
    const emptyProb = 0.7;
    const otherProb = (1 - emptyProb) * 0.7;
    const otherProbHigh = (1 - emptyProb) * 0.3;
    return weightedChoices([0, 1, 2], [emptyProb, otherProb, otherProbHigh], Config.FOREST_MAX_TIME);
  }

  static generateLeaves(): number[] {
    const arr = weightedChoices([1, 2], [0.5, 0.5], Config.FOREST_MAX_TIME);
    for (let i = 0; i < arr.length - 1; i++) if (arr[i] === 2) arr[i + 1] = 0;
    return arr;
  }
}

function weightedChoices(choices: number[], weights: number[], k: number): number[] {
  const total = weights.reduce((a, b) => a + b, 0);
  const out: number[] = [];
  for (let i = 0; i < k; i++) {
    let r = Math.random() * total;
    for (let j = 0; j < choices.length; j++) {
      r -= weights[j];
      if (r <= 0) { out.push(choices[j]); break; }
    }
    if (out.length < i + 1) out.push(choices[choices.length - 1]);
  }
  return out;
}
